src/libmongoc/tests/test-mongoc-retryable-writes.c

```diff
@@ -568,6 +568,236 @@ test_unsupported_storage_engine_error (void)
+   bool configure_second_fail;
+   char *failCommand;
+} prose_test_3_apm_ctx_t;
+
+static void
+prose_test_3_on_command_success (const mongoc_apm_command_succeeded_t *event)
+{
+   bson_iter_t iter;
+   bson_error_t error;
+   const bson_t *reply = mongoc_apm_command_succeeded_get_reply (event);
+   prose_test_3_apm_ctx_t *ctx =
+      mongoc_apm_command_succeeded_get_context (event);
+
+   // wait for a writeConcernError and then set a second failpoint
+   if (bson_iter_init_find (&iter, reply, "writeConcernError") &&
+       ctx->configure_second_fail) {
+      ctx->configure_second_fail = false;
+      ASSERT_OR_PRINT (
+         mongoc_client_command_simple (
+            ctx->client,
+            "admin",
+            tmp_bson (
+               "{'configureFailPoint': 'failCommand', 'mode': {'times': 1},"
+               " 'data': { 'failCommands': ['%s'], 'errorCode': "
+               "10107, "
+               "'errorLabels': ['RetryableWriteError', 'NoWritesPerformed']}}",
+               ctx->failCommand),
+            NULL,
+            NULL,
+            &error),
+         error);
+   }
+}
+
+uint32_t
+set_up_original_error_test (mongoc_apm_callbacks_t *callbacks,
+                            prose_test_3_apm_ctx_t *apm_ctx,
+                            char *failCommand,
+                            mongoc_client_t *client)
+{
+   uint32_t server_id;
+   bson_error_t error;
+   // clean up in case a previous test aborted
+   server_id = mongoc_topology_select_server_id (
+      client->topology, MONGOC_SS_WRITE, NULL, NULL, &error);
+   ASSERT_OR_PRINT (server_id, error);
+   deactivate_fail_points (client, server_id);
+
+   // set up callbacks for command monitoring
+   apm_ctx->client = client;
+   apm_ctx->failCommand = failCommand;
+   apm_ctx->configure_second_fail = true;
+   mongoc_apm_set_command_succeeded_cb (callbacks,
+                                        prose_test_3_on_command_success);
+   mongoc_client_set_apm_callbacks (client, callbacks, apm_ctx);
+
+   // configure the first fail point
+   bool ret = mongoc_client_command_simple (
+      client,
+      "admin",
+      tmp_bson ("{'configureFailPoint': 'failCommand', 'mode': {'times': 1}, "
+                "'data': {'failCommands': ['%s'], 'writeConcernError': {"
+                "'code': 91, 'errorLabels': ['RetryableWriteError']}}}",
+                failCommand),
+      NULL,
+      NULL,
+      &error);
+   ASSERT_OR_PRINT (ret, error);
+
+   return server_id;
+}
+
+void
+cleanup_original_error_test (mongoc_client_t *client,
+                             uint32_t server_id,
+                             bson_t *reply,
+                             mongoc_collection_t *coll,
+                             mongoc_apm_callbacks_t *callbacks)
+{
+   deactivate_fail_points (client, server_id); // disable the fail point
+   bson_destroy (reply);
+   mongoc_client_destroy (client);
+   mongoc_collection_destroy (coll);
+   mongoc_apm_callbacks_destroy (callbacks);
```

Comment:
```suggestion
   mongoc_collection_destroy (coll);
   mongoc_apm_callbacks_destroy (callbacks);
   mongoc_client_destroy (client);
```

Suggest destroying the `client` after the `coll`. `coll` is created from `client`. 

http://mongoc.org/libmongoc/current/lifecycle.html#databases-collections-and-related-objects documents the expected lifetimes:

> You can create a [mongoc_database_t](http://mongoc.org/libmongoc/current/mongoc_database_t.html) or [mongoc_collection_t](http://mongoc.org/libmongoc/current/mongoc_collection_t.html) from a [mongoc_client_t](http://mongoc.org/libmongoc/current/mongoc_client_t.html), and create a [mongoc_cursor_t](http://mongoc.org/libmongoc/current/mongoc_cursor_t.html) or [mongoc_bulk_operation_t](http://mongoc.org/libmongoc/current/mongoc_bulk_operation_t.html) from a [mongoc_collection_t](http://mongoc.org/libmongoc/current/mongoc_collection_t.html).

> Each of these objects must be destroyed before the client they were created from, but their lifetimes are otherwise independent.